Sanitize ANSII control characters returned from the server

samcoe opened the pull request:
Fixes https://github.com/github/cli/issues/149
Fixes https://github.com/cli/cli/issues/932

samcoe committed:
Sanitize ANSII control characters returned from the server

mislav commented on api/http_client.go:
```diff
@@ -115,6 +121,33 @@ func ExtractHeader(name string, dest *string) func(http.RoundTripper) http.Round
+// the values of \u0080 to \u009F represent C1 ASCII control characters. These control
+// characters will be interpreted by the terminal, this behaviour can be used maliciously
+// as an attack vector, especially the control character \u001B. This function escapes
+// all non-printable characters between \u0000 and \u00FF so that the terminal will
```
1. My understanding is that non-printable ASCII characters are in the range \u0000 and \u001F. If we escape every character up until \u00FF, wouldn't that also affect all printable characters in ASCII and beyond?
2. Would we explicitly allow-list some control ASCII characters such as LF and CR so that they don't get mangled?

samcoe replied:
1. The server only returns `\u` string sequence for non-printable characters so we do not need to worry about escaping printable characters.
2. LF and CR are printable characters and the server does not return them as `\u` sequences.

mislav commented on api/http_client.go:
```diff
@@ -64,6 +68,8 @@ func NewHTTPClient(opts HTTPClientOptions) (*http.Client, error) {
 		client.Transport = AddAuthTokenHeader(client.Transport, opts.Config)
 	}
 
+	client.Transport = SanitizeASCIIControlCharacters(client.Transport)
```
Does this middleware also affect the `gh api` command? Since `gh api` is meant for low-level, programmatic use, maybe it would be good if it remained unaffected.

samcoe replied:
It does currently affect the `gh api` command. In my mind I think we should sanitize those requests as well as we want to protect our users throughout our tool. I can envision people copying and pasting `gh api` commands that request a URL with malicious data in the response. If there is a valid use case for leaving these requests un-sanitized then I am open to it but I have been unable to think of one.

mislav replied:
The way I see it, `gh api` is meant for scripts, not necessarily to implement presentation. But agreed that flags like `--jq` and especially `--template` can be used to format data for the terminal. In those cases it's good if ANSI escape codes are disabled. Maybe it's okay that this middleware sanitizes all those.

mislav reviewed (commented):
Looks quite elegant! Thanks!

samcoe committed:
New approach

mislav commented on api/http_client.go:
```diff
@@ -115,6 +122,192 @@ func ExtractHeader(name string, dest *string) func(http.RoundTripper) http.Round
+			res.ContentLength == 0 {
+			return res, err
+		}
+		var sanitized bytes.Buffer
+		err = replaceControlCharacters(res.Body, &sanitized)
+		if err != nil {
+			err = fmt.Errorf("ascii control characters sanitization error: %w", err)
+		}
+		res.Body.Close()
+		res.Body = io.NopCloser(&sanitized)
```
One consequence of this middleware is that by default, all Bodies of every response will first be fully read into a Buffer before the middleware yields control to other middleware or to other response processing code. This prevents any kind of streaming processing of response JSON.

Would it be feasible to replace `resp.Body` with an `io.ReadCloser` that does escaping on the fly as it forwards reads to the original response body and returns each result? That way, every response from the server will be immediately available to the code that makes the request without this middleware potentially delaying that.

samcoe replied:
I like that idea. I will investigate implementing it in that manner.

mislav commented on api/http_client.go:
```diff
@@ -115,6 +122,192 @@ func ExtractHeader(name string, dest *string) func(http.RoundTripper) http.Round
+// mapControlCharacterToCaret maps C0 control sequences to caret notation and
+// C1 control sequences to hex notation. C1 control sequences do
+// not have caret notation representation.
+func mapControlCharacterToCaret(b []byte) ([]byte, bool) {
```
Love the bit of added UX by mapping unsafe escape codes to their printable equivalents 👍

mislav commented:
Forgot one comment in my last review: could there be a test that confirms that this middleware will not try to escape already escaped unicode codepoints in JSON? E.g. that `\\u001b` should be allowed to stay `\\u001b` without being converted to `\\\u001b`

samcoe committed:
Streaming and also move to new file for clarity

samcoe commented:
@mislav I changed the implementation to allow for streaming JSON responses, and I also moved the sanitization code to its own files for clarity. 

The test I have does include some escaped sequences as well, let me know if you want me to make it more explicit.

mislav commented on api/sanitize_ascii.go:
```diff
@@ -0,0 +1,196 @@
+			break
+		}
+
+		if win[0] == 92 && win[1] == 117 && win[2] == 48 && win[3] == 48 {
```
Nit: for clarity, these numbers could be saved into consts and expressed using Go single quotes instead of ASCII codes. For example, `'\\'` and `'u'` should be more clear than 92 and 117, respectively.

However, in this case, maybe `bytes.HasPrefix` could suffice?
```suggestion
		if bytes.HasPrefix(win, []byte("\\u00")) {
```

mislav commented on api/sanitize_ascii.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+// Close closes the wrapped ReadCloser.
+func (s *sanitizeASCIIReadCloser) Close() error {
```
Nit: since your struct is embedding an `io.ReadCloser` interface, I don't think you need to explicitly declare this method. Your type should already automatically have a `Close()` method and forward it to the underlying ReadCloser.

samcoe replied:
I gave the `ReadCloser` wrapped by `sanitizeASCIIReadCloser` a name in the struct so Go won't promote the `Close` method to be on `sanitizeASCIIReadCloser`. Having said that I think it makes sense to making the `ReadCloser` actually embedded.

mislav reviewed (approved):
This is a great improvement; thank you! Only style nits remain

samcoe committed:
Clean up nits
